### DISCUSSION TITLE
C2PA / CAI?

### DISCUSSION THREAD

**Discussion by carlseibert3** [ImageMagick/ImageMagick#6952] [Contribute]
### ImageMagick version

7.1.1

### Operating system, version and so on

all

### Description

Hello,  Has there been any discussion here about the Content Authenticity Initiative / C2PA process for verifying image authenticity? Search didn't surface anything. C2PA has just become a real concern/possibility/opportunity with the release of cameras capable of following the standard. 

In a nutshell, C2PA metadata offers a secure method for anyone who wants to establish that a given image matches the original from a camera or phone. The principle intent is to combat disinformation, but other potential use cases include things like photos for insurance claims or evidence of just about anything.

The process creates signed thumbnails of an image at creation and after editing and other steps in a publication workflow.  

In order to publish a signed image, a web CMS would have to create the various-sized renditions necessary for responsive pages without breaking the signature chain of said image.

That would mean creating and signing what C2PA calls a "manifest" and attaching it - or a pointer to it - to the image after resizing it.

ImageMagick is used by CMSes like WordPress, Drupal, and thousands more to do that resizing. When website operators tumble to the notion that having C2PA- signed images differentiates legitimate sites concerned with their reputation from those that aren't, they're going to be all in a panic to be able to use C2PA.  They'll be all in a snit wanting ImageMagick to develop something non-trivial overnight.

Thus, I'm writing now to give you a heads-up.

I just posted a long post on my that attempts to explain the process in some detail. It's meant for photographers, not developers, but there may be some information of use there.  https://www.carlseibert.com/first-camera-with-cai-content-credentials-introduced/

If I can be of any help, please feel free to reach out.

**Comment by snibgo** [comment]
Thanks for your interesting and informative article.

I think adding C2PA capability to ImageMagick (IM) would be a good thing, in principle. Describing all image-processing operations in unhackable* metadata would be neat, but would need massive changes to IM. An easier change would be to add an option to create unhackable metadata of some text or thumbnails.

I don't understand how this would prevent a hostile actor from adding false metadata. IM is open source. Anyone can download the source code, alter that in any way, and rebuild IM. So anyone could publish an image that claimed to have had only innocent editing (eg adjusting lightness) when in fact objects had been added and subtracted.

*Unhackable: I mean, encoded cryptographically, so it can't be modified without massive difficulty.

**Reply by carlseibert3** [reply]
> " create unhackable metadata of some text or thumbnails"   That's exactly what it does. The verbal descriptions of what was done to the image are very brief and general - "tonal adjustments", it cropping adjustments".  Kind of a sideshow. The thumbnails that it makes at each stage of the image's life are the main attraction. 

The C2PA process doesn't prevent bad actors from acting, it just makes the image tamper-evident.  Adding new metadata, false or otherwise, would break the trust chain if it's done outside of a C2PA-enabled app, thus altering the world to the fact that the image had been changed. 

On the other hand, if the original metadata had been included in a C2PA manifest, either embedded or cloud-based, then it would be available to whoever might look at the image with a C2PA verification tool later. That's kind of the foundational principle of the thing: downstream users can see the original content.  (Of course, I'm talking about IPTC metadata here. The caption and authorship information mainly.)

**Comment by urban-warrior** [comment]
ImageMagick currently offers several methods to provide some level of authenticity to an image:

1. Image signature, e.g., `identify -verbose logo: | grep -i Sign`.
2. Encrypted pixels with a secret key, e.g., `-encipher, -deciper`.
3. And steganography, e.g., `-stegano`.

Outside of ImageMagick, there is no reason why images could not be signed in the same way we sign code releases.

Next, let's consider C2PA. Due to our limited development team, our strategy involves waiting for an open-source library to handle the intricate aspects of C2PA. ImageMagick would subsequently integrate this library through a few fundamental API calls to leverage its capabilities. Unless a user provides a solution, it is highly improbable that ImageMagick will support C2PA in the interim. Even if a user does contribute a solution, it must remain an optional feature to prevent code bloat. Additionally, it should encapsulate the majority of the effort within a separate library with its own API, effectively minimizing the impact on ImageMagick's footprint.

**Reply by carlseibert3** [reply]
I think the good news here is that the CAI and C2PA seem to be releasing open-source tools. So, that code should be helpful. I have no idea when the outlets that will want to provide C2PA support in their products will tumble to the fact that most of them use ImageMgick in their CMSes to make their responsive renditions. And I have no real idea whether the most practical approach will be to build something that uses ImageMagick as it is, as a library within a larger application or to work with you to build the functionality into ImageMagick itself. On the good news side, many of these organizations have considerable development resources to contribute. On the bad news side is that we can be well assured that they will wait until the last minute and then panic. (I was in the business for a long time.)   So, forewarned is fore-armed or whatever it is that people say. Considering the implications for society, proactive would be the best stance.

**Comment by bevand10** [comment]
To @carlseibert3 's point, there is open-source software available from the C2PA/CAI consortium to perform the signing action.

The library has been produced as a Rust crate, and explicitly designed allow integration into tools - be that WASM in the browser, to static binaries provided by application providers such as the ImageMagick project.

The c2pa team provide this library for C programming teams: https://opensource.contentauthenticity.org/docs/c2pa-c/
